### PR TITLE
STM-based chain selection algo for mining

### PR DESCRIPTION
**Pull request by edmundnoble**

### CODE REVIEW

File: src/Chainweb/Miner/Coordinator.hs [hunk 1]
```diff
@@ -298,43 +294,83 @@ work
...
+    T2 wh pd <-
+        withAsync (logDelays 0) $ \_ -> findWork
     now <- getCurrentTimeIntegral
     atomically
         . modifyTVar' (_coordState mr)
         . over miningState
         . M.insert (_payloadDataPayloadHash pd)
         $ T3 m pd now
-    return wh
+    pure wh
```
Comment (chessai) [inline src/Chainweb/Miner/Coordinator.hs:306]: let's replicate this everywhere

**Review by chessai — APPROVED** [review]

File: src/Chainweb/Miner/Coordinator.hs [hunk 2]
```diff
@@ -298,43 +294,83 @@ work
...
+        logf @T.Text Info $ case HM.lookup (view minerId m) primedWork of
+            Nothing ->
+                 "findWork: no primed work for miner key" <> sshow m
+            Just mpw
+                | HM.null mpw ->
+                    "findWork: no chains have primed work"
+                | otherwise ->
+                    "findWork: all chains with primed work may be stalled, possible stalled chains: " <> sshow (HM.keys mpw)
+
+        logDelays (n + 1)
```
Comment (chessai) [inline src/Chainweb/Miner/Coordinator.hs:336]:
```suggestion
        logDelays n'
```
you already define this above, so why not

File: src/Chainweb/Miner/Coordinator.hs [hunk 3]
```diff
@@ -164,68 +169,59 @@ newtype PrevTime = PrevTime BlockCreationTime
...
-chainChoice c choice = case choice of
-    Anything -> randomChainIdAt c (minChainHeight c)
-    Suggestion cid -> pure cid
-    TriedLast cid -> loop cid
-  where
-    loop :: ChainId -> IO ChainId
-    loop cid = do
-        new <- randomChainIdAt c (minChainHeight c)
-        bool (pure new) (loop cid) $ new == cid
+-- -- | Construct a new `BlockHeader` to mine on.
```
Comment (chessai) [inline src/Chainweb/Miner/Coordinator.hs:172]: pls delet this

File: src/Chainweb/Cut/Create.hs [hunk 4]
```diff
@@ -213,13 +213,13 @@ getCutExtension c cid = do
     tryAdj b
 
         -- When the block is behind, we can move ahead
-        | _blockHeight b == targetHeight = Just $! _blockParent b
+        | _blockHeight b == targetHeight = Just (_blockParent b)
```
Comment (chessai) [inline src/Chainweb/Cut/Create.hs:216]: undo pls. selector thunk

File: src/Chainweb/Cut/Create.hs [hunk 5]
```diff
@@ -213,13 +213,13 @@ getCutExtension c cid = do
...
 
         -- if the block is ahead it's blocked
         | _blockHeight b + 1 == parentHeight = Nothing -- chain is blocked
 
         -- If this is not a graph transition cut we can move ahead
-        | _blockHeight b == parentHeight = Just $! _blockHash b
+        | _blockHeight b == parentHeight = Just (_blockHash b)
```
Comment (chessai) [inline src/Chainweb/Cut/Create.hs:222]: same thing

**Comment by edmundnoble** [comment]
Closed in favor of https://github.com/kadena-io/chainweb-node/pull/1758 to avoid potential negative performance implications.

**Comment by chessai** [comment]
> Closed in favor of https://github.com/kadena-io/chainweb-node/pull/1758 to avoid potential negative performance implications.

this is #1758

**Pull request closed without merging**